### PR TITLE
Trap convenience class

### PR DESCRIPTION
**Pull request by jhdark**
## Proposed changes

With this update users will be able to give a trap object without having to define a 2 species (immobile and implicit) and a reaction, instead just having a `F.Trap()` object instead and passing it to the new attribute of `F.HydrogenTransporModel`, `traps`. Traps needs to be given as a list like so:

```python
my_trap = F.Trap(
    name="test",
    species=my_mobile_species,
    k_0=1,
    E_k=1,
    p_0=1,
    E_p=1,
    n=1,
    volume=my_vol,
)
my_model.traps = [my_trap]
```

This convenience class should make some simpler hydrogen trapping cases easier for users

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jhdark** [comment]
@RemDelaporteMathurin for some reason when the final test I've added in `test/test_h_transport_problem` breaks everything else, but when its commented the rest of the tests pass, not sure whats going on here?

**Comment by RemDelaporteMathurin** [comment]
@jhdark I figured out the issue. It is related to the fact that the default value for `reactions` in `HTransportProblem` is `[]`.

When adding a trap object (like in the test), [on this line](https://github.com/jhdark/FESTIM/blob/9caaf122c398fbc4abdad2a5196d7d95e4b94164/festim/hydrogen_transport_problem.py#L222), a reaction is added to self.reactions. However, since `self.reactions` is a list and lists in python are class instances, the (unwanted) result is that a reaction is added to the default value of `self.reactions`.

Consider this where we first create a model with a dummy reaction, then another empty model, and the reaction is carried over.

```python
import festim as F

model1 = F.HydrogenTransportProblem()
model1.reactions.append("test_reaction_as_str")


model2 = F.HydrogenTransportProblem()
print(model2.reactions)  # IT SHOULD BE EMPTY!!
```

**Comment by RemDelaporteMathurin** [comment]
@gabriele-ferrero @SamueleMeschini would you mind having a look at this, at least from the user point of view? Is this implementation fine with you? The idea is to later have even more convenience by having one default mobile species created if nothing is given by the users.

File: festim/hydrogen_transport_problem.py [hunk 1]
```diff
@@ -190,6 +194,10 @@ def species(self, value):
         self._species = value
 
     def initialise(self):
+        # if traps are given, create the necessary objects
+        if len(self.traps) >= 1:
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:198]:
```suggestion
        if len(self.traps) >0:
```

File: festim/species.py [hunk 2]
```diff
@@ -62,23 +63,60 @@ class Trap(Species):
 
     Args:
         name (str, optional): a name given to the trap. Defaults to None.
+        species (F.Species): the mobile species to be trapped
```
Comment (RemDelaporteMathurin) [inline festim/species.py:66]: Should we call this `mobile_species` to be more explicit?

File: festim/species.py [hunk 3]
```diff
@@ -62,23 +63,60 @@ class Trap(Species):
...
+        k_0 (float): the trapping rate constant pre-exponential factor.
+        E_k (float): the trapping rate constant activation energy.
+        p_0 (float): the detrapping rate constant pre-exponential factor.
+        E_p (float): the detrapping rate constant activation energy.
```
Comment (RemDelaporteMathurin) [inline festim/species.py:70]: Maybe we should give units

File: festim/species.py
Comment (RemDelaporteMathurin) [inline festim/species.py:1]: The `Trap` class would benefit from more precise documentation. Like specifying this only works for 1 mobile species, only one trapping level, etc.

Maybe even giving the equivalence with explicitly giving the species and reactions

File: test/test_h_transport_problem.py [hunk 4]
```diff
@@ -645,3 +645,25 @@ def test_species_setter():
...
+        p_0=1,
+        E_p=1,
+        n=1,
+        volume=my_vol,
+    )
+    my_model.traps = [my_trap]
+    my_model.create_species_from_trap()
+
+    assert len(my_model.species) == 1
+    assert len(my_model.reactions) == 1
```
Comment (RemDelaporteMathurin) [inline test/test_h_transport_problem.py:669]: I would also test that the element in `my_model.species` and `my_model.reactions` are the correct type.

These tests would pass if we added `"coucou"` to the list

File: test/test_species.py [hunk 5]
```diff
@@ -115,3 +115,29 @@ def test_implicit_species_concentration_with_no_solution():
     # test that a ValueError is raised when the second species has no solution
     with pytest.raises(ValueError):
         implicit_species.concentration
+
+
+def test_create_species_and_reaction():
+    """test that the trapped_concentration and trap_reaction attribites
```
Comment (RemDelaporteMathurin) [inline test/test_species.py:121]:
```suggestion
    """test that the trapped_concentration and trap_reaction attributes
```

File: test/test_h_transport_problem.py (hunk 4 above)
Comment (RemDelaporteMathurin) [reply test/test_h_transport_problem.py:669]: I see you did it in the other test. Nonetheless, it isn't hard to add a type check here too

**Comment by jhdark** [comment]
> > @gabriele-ferrero @SamueleMeschini would you mind having a look at this, at least from the user point of view? Is this implementation fine with you? The idea is to later have even more convenience by having one default mobile species created if nothing is given by the users.
> 
> It seems ok, the notes do a great job explaining the equivalence of the convinience class. A default mobile species would be even better because I think most of the users will do simple H transport.

Yeah a default mobile species will be needed, its just to have a default mobile species, this would require refactoring of a lot of other logic in `hydrogen_transport_problem`, so think we'll have to save that for another PR

**Comment by gabriele-ferrero** [comment]
It seems ok to me, and I think it is a great convenience class to have in festim. I will have a further look when it will pass the checks

**Comment by RemDelaporteMathurin** [comment]
> when it will pass the checks

Yes this should be fixed in #661

File: festim/hydrogen_transport_problem.py [hunk 6]
```diff
@@ -208,6 +216,14 @@ def initialise(self):
         self.create_solver()
         self.initialise_exports()
 
+    def create_species_from_trap(self):
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:219]: Should we have traps (plural) here instead?

File: festim/species.py [hunk 7]
```diff
@@ -60,25 +61,104 @@ def concentration(self):
...
+        self.E_p = E_p
+        self.n = n
+        self.volume = volume
+
+    def create_species_and_reaction(self):
+        """create the immobile trapped species object and the reaction for trapping"""
+        self.trapped_concentration = F.Species(name=self.name, mobile=False)
+        trap_site = F.ImplicitSpecies(n=self.n, others=[self.trapped_concentration])
+
+        self.trap_reaction = F.Reaction(
```
Comment (RemDelaporteMathurin) [inline festim/species.py:152]: I would just call this `reaction`

File: festim/species.py [hunk 8]
```diff
@@ -60,25 +61,104 @@ def concentration(self):
...
+    ) -> None:
         super().__init__(name)
+        self.mobile_species = mobile_species
+        self.k_0 = k_0
+        self.E_k = E_k
+        self.p_0 = p_0
+        self.E_p = E_p
+        self.n = n
+        self.volume = volume
+
```
Comment (RemDelaporteMathurin) [inline festim/species.py:146]: Let's initialise all the attributes to `None` here (reaction, trapped_concentration, etc.)

File: festim/hydrogen_transport_problem.py [hunk 9]
```diff
@@ -193,6 +197,10 @@ def species(self, value):
         self._species = value
 
     def initialise(self):
+        # if traps are given, create the necessary objects
+        if len(self.traps) > 0:
```
[the remark this answers is not in the document]
Comment (jhdark) [reply festim/hydrogen_transport_problem.py:201]: Yeah think we can get rid of this

File: festim/hydrogen_transport_problem.py (hunk 6 above)
Comment (jhdark) [reply festim/hydrogen_transport_problem.py:219]: sure

File: festim/species.py (hunk 7 above)
Comment (jhdark) [reply festim/species.py:152]: sure

**Pull request merged**